src/mbgl/renderer/render_layer.hpp

```diff
@@ -65,6 +65,9 @@ class RenderLayer {
     // Returns true if the layer has a pattern property and is actively crossfading.
     virtual bool hasCrossfade() const = 0;
 
+    // Returns true if layer writes to depth buffer by drawing using PaintParameters::depthModeFor3D().
+    virtual bool is3D() const { return false; }
```

Comment:
No. Pass3D is used for offscreen render of heatmap and hillshade and should be, at least renamed to Offscreen. It is a relict of previous implementation and not used anymore [since fill extrusion is done during Translucent pass](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/renderer/layers/render_fill_extrusion_layer.cpp#L59). is3D is only to be true for fill extrusion, not for heatmap or hillshade.

Reply:
Oh, this naming is confusing indeed :-/ maybe it's worth introducing another `LayerTypeInfo::` flag than to be enabled for fill-extrusion ?
:+1: for renaming `LayerTypeInfo::Pass3D ` to `LayerTypeInfo::Offscreen` (not necessary in this PR)

Reply:
Flag is not needed. I'd like to follow approach in mapbox-gl-js: note that we still need to implement similar for [customLayer::is3D()](https://github.com/mapbox/mapbox-gl-js/blob/8bcf115473a9fcd631cc283685b5d4a50c141522/src/style/style_layer/custom_style_layer.js#L195).

Yes, renaming passes is better not to have here. It is orthogonal.

Reply:
Makes sense, thanks for clarification.